index.js

```diff
@@ -23,24 +20,24 @@ export default (sbp('sbp/selectors/register', {
       eventQueues[name] = { events: [] }
     }
     const events = eventQueues[name].events
-    events.push({
+    const thisEvent = {
       sbpInvocation,
-      state: STATE_PENDING,
       promise: null
-    })
+    }
+    events.push(thisEvent)
     while (events.length > 0) {
       const event = events[0]
-      if (event.state === STATE_PENDING) {
-        event.state = STATE_INVOKING
-        event.promise = sbp(...event.sbpInvocation)
-        await event.promise
-        event.state = STATE_FINISHED
-      } else if (event.state === STATE_INVOKING) {
+      if (event === thisEvent) {
+        const promise = sbp(...event.sbpInvocation)
+        event.promise = new Promise((accept) => promise.finally(accept))
```

Comment:
It's unclear to me why this is needed or what this is even doing. It doesn't seem to make any difference afaict (except for other issues noted in review)

Reply:
It shouldn't break anything, because subsequent calls wait on `event.promise` to be finished. The state is a bit redundant, as `await` alone in this case can do the waiting.

For example:

1. Empty queue: `event === thisEvent`, so it will be executed.
2. Non-empty queue: `else` will be executed, and it will wait on the first event to be done. It will continue to do so until the event is first on the queue, then the `event === thisEvent` branch runs

Reply:
On the other hand, this fixes a few issues with the original code:

1. Inability to access return values
2. Not handling errors

Reply:
`event.promise = new Promise((accept) => promise.finally(accept))`

This creates a promise that will be done once `promise` is finished (whether it failed or succeeded). It's not strictly needed (alternative is `event.promise = promise`), but it has two additional advantages:

1. No need to handle the fail case in the `else` below
2. It encapsulates the original promise, so that it's only available to the calling scope.

Reply:
Hmm, I will need to think on it some more. There does seem to be some nice simplification here.

However, something that I think is a real issue is:

> 2. Not handling errors

That's a feature not a bug :P

There is a lot of code in the project that that relies on this. For example, `'chelonia/contract/sync'` and `'chelonia/contract/remove'` both should throw exceptions that can be caught by the caller if any of their inner queued events fail.

I'm not completely sure, but it seems like this PR might break that?

Reply:
This does throw an exception in that case. The issue with the original code is that the same error could appear in different places.

For example, if you have a queue with events `A`, `B`, `C` and `A` fails, I believe the calls for all three would fail (originally). This only lets `A` fail, while `B` and `C` are treated as independent. Also, as far as I understand, the queue would be left forever unusable in the previous code, as a failing event would never be cleared up.